crypto/s2n_aead_cipher_aes_gcm.c

```diff
@@ -23,26 +23,126 @@
 #include "utils/s2n_safety.h"
 #include "utils/s2n_blob.h"
 
+#if defined(OPENSSL_IS_BORINGSSL) || defined(OPENSSL_IS_AWSLC)
+#define S2N_AEAD_AES_GCM_AVAILABLE_BSSL_AWSLC
+#endif
+
 static uint8_t s2n_aead_cipher_aes128_gcm_available()
 {
+#if defined(S2N_AEAD_AES_GCM_AVAILABLE_BSSL_AWSLC)
+    return (EVP_aead_aes_128_gcm() ? 1 : 0);
+#else
     return (EVP_aes_128_gcm() ? 1 : 0);
+#endif
 }
 
 static uint8_t s2n_aead_cipher_aes256_gcm_available()
 {
+#if defined(S2N_AEAD_AES_GCM_AVAILABLE_BSSL_AWSLC)
+    return (EVP_aead_aes_256_gcm() ? 1 : 0);
+#else
     return (EVP_aes_256_gcm() ? 1 : 0);
+#endif
+}
+
+#if defined(S2N_AEAD_AES_GCM_AVAILABLE_BSSL_AWSLC) /* BoringSSL and AWS-LC AEAD API implementation */
+
+static int s2n_aead_cipher_aes_gcm_encrypt(struct s2n_session_key *key, struct s2n_blob *iv, struct s2n_blob *aad, struct s2n_blob *in, struct s2n_blob *out)
+{
+    /* The size of the |in| blob includes the size of the data and the size of the AES-GCM tag */
+    gte_check(in->size, S2N_TLS_GCM_TAG_LEN);
+    gte_check(out->size, in->size);
+    eq_check(iv->size, S2N_TLS_GCM_IV_LEN);
```

Comment:
Do you need notnull checks for in/out/iv/key/aad?

Reply:
You could dereference NULL here. So, in principal yes. But it would break the pattern used in all the other functions. My guess is it was implicitly assumed that, at this layer, all input is not NULL, to simplify the code.

Reply:
We've had issues before because a layer assumed input couldn't be NULL, but then the layer above it silently changed. I'd add the checks to your new methods just in case, even if you don't want to update any of the existing methods.

Reply:
Added null-checks to new aead functions.